In [134]:
# Import required libraries and dependencies
import pandas as pd
import numpy as np
import hvplot.pandas
from pathlib import Path
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [135]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [136]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [137]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [138]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
market_data_scaled = StandardScaler().fit_transform(
    df_market_data[["price_change_percentage_24h", "price_change_percentage_24h", "price_change_percentage_7d", "price_change_percentage_30d", "price_change_percentage_60d", "price_change_percentage_200d", "price_change_percentage_1y"]]
)

# Review the scaled data
market_data_scaled [0:5]

array([[ 0.50852937,  0.50852937,  0.49319307,  0.23545963, -0.0674951 ,
        -0.35595348, -0.25163688],
       [ 0.18544589,  0.18544589,  0.93444504, -0.05434093, -0.27348273,
        -0.11575947, -0.19935211],
       [ 0.02177396,  0.02177396, -0.70633685, -0.06103015,  0.00800452,
        -0.55024692, -0.28206051],
       [-0.04076438, -0.04076438, -0.81092807, -0.05038797, -0.37316402,
        -0.45825882, -0.29554614],
       [ 1.19303608,  1.19303608,  2.00095907,  0.54584206, -0.29120287,
        -0.49984776, -0.27031695]])

In [139]:
# Create a DataFrame with the scaled data
market_data_scaled_df = pd.DataFrame(
    market_data_scaled,
    columns=["price_change_percentage_24h", "price_change_percentage_7d", "price_change_percentage_14d", "price_change_percentage_30d", "price_change_percentage_60d", "price_change_percentage_200d", "price_change_percentage_1y"]
)


# Copy the crypto names from the original data
market_data_scaled_df["coin_id"] = df_market_data.index

# Set the coinid column as index
market_data_scaled_df = market_data_scaled_df.set_index("coin_id")


# Display sample data
market_data_scaled_df.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.508529   
ethereum                         0.185446                    0.185446   
tether                           0.021774                    0.021774   
ripple                          -0.040764                   -0.040764   
bitcoin-cash                     1.193036                    1.193036   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.493193                     0.235460   
ethereum                         0.934445                    -0.054341   
tether                          -0.706337                    -0.061030   
ripple                          -0.810928                    -0.050388   
bitcoin-cash                     2.000959                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Data.

In [140]:
# Create a list with the number of k-values from 1 to 11
k = list (range(1,11))

In [141]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list
for i in k:
    k_model = KMeans(n_clusters=i, random_state=0)
    k_model.fit(market_data_scaled_df)
    inertia.append(k_model.inertia_)

/Users/ugochiprinceajuobi/anaconda3/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
/Users/ugochiprinceajuobi/anaconda3/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
/Users/ugochiprinceajuobi/anaconda3/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
/Users/ugochiprinceajuobi/anaconda3/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppre

In [142]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {"k": k, "inertia": inertia}

# Create a DataFrame with the data to plot the Elbow curve
df_elbow = pd.DataFrame(elbow_data)

# Review the DataFrame
df_elbow.head()

k     inertia
0  1  287.000000
1  2  174.633932
2  3  106.871252
3  4   74.302710
4  5   59.238794

In [143]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
df_elbow.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve using Scaled Market Data ", 
    xticks=k
)

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** From this curve I observed the the best value for 'k'to be 3 

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [144]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=3, random_state=0)

In [145]:
# Fit the K-Means model using the scaled data
model.fit(market_data_scaled_df)

/Users/ugochiprinceajuobi/anaconda3/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


KMeans(n_clusters=3, random_state=0)

In [146]:
# Predict the clusters to group the cryptocurrencies using the scaled data
market_data_scaled_clusters = model.predict(market_data_scaled_df)

# Print the resulting array of cluster values.
print(market_data_scaled_clusters)

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 2
 0 0 0 0]


In [147]:
# Create a copy of the DataFrame
market_scaled_predictions_df = market_data_scaled_df.copy()

In [148]:
# Add a new column to the DataFrame with the predicted clusters
market_scaled_predictions_df["market_data_scaled_cluster"] = market_data_scaled_clusters

# Display sample data
market_scaled_predictions_df.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.508529   
ethereum                         0.185446                    0.185446   
tether                           0.021774                    0.021774   
ripple                          -0.040764                   -0.040764   
bitcoin-cash                     1.193036                    1.193036   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.493193                     0.235460   
ethereum                         0.934445                    -0.054341   
tether                          -0.706337                    -0.061030   
ripple                          -0.810928                    -0.050388   
bitcoin-cash                     2.000959                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  market_data_scaled_cluster  
coin_id                                                               
bitcoin                        -0.251637                           0  
ethereum                       -0.199352                           0  
tether                         -0.282061                           0  
ripple                         -0.295546                           0  
bitcoin-cash                   -0.270317                           0

In [149]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
# Color by cluster labels
# Show cryptocurrency names on hover

market_scaled_predictions_df.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    by="market_data_scaled_cluster",  
    hover_cols= ["coin_id"],
    title = "Scatter Plot by Market Cluster - k=3")
 

:NdOverlay   [market_data_scaled_cluster]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [150]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)

In [151]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.
market_data_pca = pca.fit_transform(market_data_scaled_df)

# View the first five rows of the DataFrame. 
market_data_pca[:5]

array([[-0.86892187,  0.18737855,  0.44104455],
       [-0.45619773, -0.18365748,  0.90546579],
       [-0.34276986, -0.22580745, -0.79697545],
       [-0.26522949, -0.45817232, -0.83535108],
       [-1.81673834,  0.48719503,  1.96064463]])

In [152]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
pca.explained_variance_ratio_

array([0.46911654, 0.30315354, 0.14581802])

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:**
Total Explained Variance of the three PCAs = 0.46911654 + 0.30315354 + 0.14581802

Total Explained Variance ≈ 0.9180881 (rounded to a reasonable number of decimal places)

In [153]:
# Create a new DataFrame with the PCA data.
# Creating a DataFrame with the PCA data
df_market_data_pca = pd.DataFrame(market_data_pca, columns=["PCA1", "PCA2", "PCA3"])

# Copy the crypto names from the original data
df_market_data_pca["coin_id"] = df_market_data.index


# Set the coinid column as index
df_market_data_pca = df_market_data_pca.set_index("coin_id")


# Display sample data
df_market_data_pca.head()

PCA1      PCA2      PCA3
coin_id                                   
bitcoin      -0.868922  0.187379  0.441045
ethereum     -0.456198 -0.183657  0.905466
tether       -0.342770 -0.225807 -0.796975
ripple       -0.265229 -0.458172 -0.835351
bitcoin-cash -1.816738  0.487195  1.960645

---

### Find the Best Value for k Using the PCA Data

In [154]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1, 11))

In [155]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list
for i in k:
    k_model = KMeans(n_clusters=i, random_state=0)
    k_model.fit(df_market_data_pca)
    inertia.append(k_model.inertia_)

/Users/ugochiprinceajuobi/anaconda3/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
/Users/ugochiprinceajuobi/anaconda3/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
/Users/ugochiprinceajuobi/anaconda3/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
/Users/ugochiprinceajuobi/anaconda3/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppre

In [156]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data_pca = {"k": k, "inertia": inertia}
# Create a DataFrame with the data to plot the Elbow curve
df_elbow_pca = pd.DataFrame(elbow_data_pca)

# Review the DataFrame
df_elbow_pca.head()

k     inertia
0  1  263.491284
1  2  152.051982
2  3   84.658164
3  4   52.224796
4  5   39.168467

In [157]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
df_elbow_pca.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve using Pca Data ", 
    xticks=k
)

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** The best value for 'k'based on the curve is '3'

* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** It does not differ as both have same curve lines and best value of 'k = 3'

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [158]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=3, random_state=0)

In [159]:
# Fit the K-Means model using the PCA data
model.fit(df_market_data_pca)

/Users/ugochiprinceajuobi/anaconda3/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


KMeans(n_clusters=3, random_state=0)

In [160]:
# Predict the clusters to group the cryptocurrencies using the PCA data
pca_data_clusters = model.predict(df_market_data_pca)

# Print the resulting array of cluster values.
print(pca_data_clusters)

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 2
 0 0 0 0]


In [161]:
# Create a copy of the DataFrame with the PCA data
pca_data_predictions_df = df_market_data_pca.copy()

# Add a new column to the DataFrame with the predicted clusters
pca_data_predictions_df["pca_data_cluster"] = pca_data_clusters

# Display sample data
pca_data_predictions_df.head()



PCA1      PCA2      PCA3  pca_data_cluster
coin_id                                                     
bitcoin      -0.868922  0.187379  0.441045                 0
ethereum     -0.456198 -0.183657  0.905466                 0
tether       -0.342770 -0.225807 -0.796975                 0
ripple       -0.265229 -0.458172 -0.835351                 0
bitcoin-cash -1.816738  0.487195  1.960645                 0

In [162]:
# Create a scatter plot using hvPlot by setting 
# `x="PC1"` and `y="PC2"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
pca_data_predictions_df.hvplot.scatter(
    x="PCA1",
    y="PCA2",
    by="pca_data_cluster",  
    hover_cols= ["coin_id"],
    title = "Scatter Plot by Pca Data Cluster - k=3")
 

:NdOverlay   [pca_data_cluster]
   :Scatter   [PCA1]   (PCA2,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [30]:
# Composite plot to contrast the Elbow curves
curve_original = elbow_curve_original.hvplot.line(x='Clusters', y='Distortion', label='Original Data')
curve_pca = elbow_curve_pca.hvplot.line(x='Principal Components', y='Distortions', label='PCA Data')

combined_plot = curve_original + curve_pca

# Display the combined plot
combined_plot.opts(title="Elbow Curves for Original and PCA-Reduced Data")

In [31]:
# Composite plot to contrast the clusters
# YOUR CODE HERE!

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** 